File: programmers/2020_카카오_인턴/2-수식최대화.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;
vector<vector<int>> vv_priority = {
    {0, 1, 2},
    {0, 2, 1},
    {1, 0, 2},
    {1, 2, 0},
    {2, 0, 1},
    {2, 1, 0},
};
map<char, int> map_opNum;

long long calculate(string expression)
{
}

class parseTree
{
};

long long
solution(string expression)
{
    map_opNum['+'] = 0;
    map_opNum['-'] = 1;
    map_opNum['*'] = 2;

    for (vector<int> v_priority : vv_priority)
    {
        long long result = calculate(expression);
    }

    long long answer = 0;
    return answer;
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    long long result = solution("100-200*300-500+20");
    cout << result << endl;
    return 0;
}

File: programmers/2020_카카오_인턴/2-수식최대화.py
ll_priority = [
    ['+', '-', '*'],
    ['+', '*', '-'],
    ['-', '+', '*'],
    ['-', '*', '+'],
    ['*', '+', '-'],
    ['*', '-', '+'],
]

class ParseTree:
    def __init__(self, expression, l_priority):
        self.l_priority = l_priority
        self.value = self.parse(expression, 0)
        
    def parse(self, expression, depth):
        if(depth >= 2):
            return eval(expression)
        
        op = self.l_priority[depth]
        expList = expression.split(op)
        resultList = []
        for exp in expList:
            result = str(self.parse(exp, depth + 1))
            resultList.append(result)
        resultExp = op.join(resultList)
        return eval(resultExp)

    def getValue(self):
        return abs(self.value)

def solution(expression):
    answer = 0
    for l_priority in ll_priority:
        parseTree = ParseTree(expression, l_priority)
        result = parseTree.getValue()
        answer = max(answer, result)
    print(answer)
    return answer
    
solution("50*6-3*2")